packages/kbn-custom-integrations/src/state_machines/services/validation.ts

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { isEmpty } from 'lodash';
+import { InvokeCreator } from 'xstate';
+import { IntegrationError } from '../../types';
+
+export class FormattingError extends IntegrationError {
+  constructor(message?: string) {
+    super(message);
+    Object.setPrototypeOf(this, new.target.prototype);
+  }
+}
+
+type Validator = (field: string) => IntegrationError | null;
+
+interface ValidatorsConfig {
+  [key: string]: Validator[] | ((arrayField: unknown[]) => IndexedValidationErrors | null);
+}
+
+export interface ValidationErrors {
+  [key: string]: ValidationResult;
+}
+
+export interface IndexedValidationErrors {
+  [key: number]: {
+    [key: string]: IntegrationError[];
+  };
+}
+
+type ValidationResult = IntegrationError[] | IndexedValidationErrors;
+
+export const initializeValidateFields =
+  (validatorsConfig: ValidatorsConfig): InvokeCreator<any, any> =>
+  (context) => {
+    const errors = validateConfigsAgainstContext(validatorsConfig, context.fields);
+    if (Object.keys(errors).length > 0) {
+      return Promise.reject({ errors });
+    } else {
+      return Promise.resolve();
+    }
+  };
+
+export const createIsEmptyValidation = (message: string) => (field: unknown) =>
+  isEmpty(field) ? new FormattingError(message) : null;
+
+export const createIsLowerCaseValidation = (message: string) => (field: string) =>
+  field.toLowerCase() !== field ? new FormattingError(message) : null;
+
+export const createCharacterLimitValidation = (message: string, limit: number) => (field: string) =>
+  field.length > limit ? new FormattingError(message) : null;
+
+export const createArrayValidator = (validatorsConfig: ValidatorsConfig) => {
+  return (arrayField: any[]) => {
+    const arrayErrors = arrayField.reduce<IndexedValidationErrors>(
+      (indexedErrors, item, currentIndex) => {
+        const errorsForField = validateConfigsAgainstContext(validatorsConfig, item);
+        return {
+          ...indexedErrors,
+          ...(Object.keys(errorsForField).length > 0 ? { [currentIndex]: errorsForField } : {}),
+        } as IndexedValidationErrors;
+      },
+      {}
+    );
+
+    return Object.keys(arrayErrors).length > 0 ? arrayErrors : null;
+  };
+};
+
+const validateConfigsAgainstContext = (validatorsConfig: ValidatorsConfig, context: any) => {
```

Comment:
**suggestion**: took me some time to get the logic of this validation and I noticed the cases for `validatorsOrValidator` being a single function or an array are handled in parallel.
We could improve its readability always converting the value into an array to have a single logical flow, wdyt?
This snippet could fit the case:
```ts
const validateConfigsAgainstContext = (validatorsConfig, context) => {
  return Object.entries(validatorsConfig).reduce((errors, [field, validatorsOrValidator]) => {
    const validators = Array.isArray(validatorsOrValidator)
      ? validatorsOrValidator
      : [validatorsOrValidator];

    const fieldErrors = validators.flatMap((validator) => {
      const error = validator(context[field]);
      return error === null ? [] : [error];
    });

    return fieldErrors.length > 0 ? { ...errors, [field]: fieldErrors } : errors;
  }, {});
};
```

Reply:
I agree your version is more readable, but the resulting data structure is slightly different (array item errors aren’t an object keyed by array index). 

I’m going to merge this PR without this change for now (just don’t want such a big PR sitting around). But I will make this more readable as part of my PR with the state machine tests 👍